### PR TITLE
Create personal account part 8

### PR DESCRIPTION
**Pull request by mejdoo**
## What's new in this PR?

Follow up of #2764 

- Implemented the logic to pass data between registration screens via a sharedViewModel.
- Created a new dependency injection module `register` to interact with the registration endpoint `/register`.

https://wearezeta.atlassian.net/browse/AN-6840

### TODO in the next PRs 

- Comparing with iOS client to Identify the missing UI/UX in the flow.
- Tuning the design of the screens
- Adding the missing validations for enter name/password fields.
- Handling no connection status/ loading status
- Verify error copy to keep consistency with iOS.
- Analyse and mimic how the scala code is adding a new user to the global database
- Connecting the last registration screen with scala 
#### APK
[Download build #1919](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1919/artifact/build/artifact/wire-dev-PR2775-1919.apk)

### CODE REVIEW

File: app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCaseTest.kt [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        registerPersonalAccountWithEmailUseCase = RegisterPersonalAccountWithEmailUseCase(activationRepository)
+    }
+
+    @Test
+    fun `Given register personal account with email use case is executed, when there is a Forbidden error then return UnauthorizedEmail`() = runBlockingTest {
+
+        `when`(registrationParams.name).thenReturn(TEST_NAME)
+        `when`(registrationParams.email).thenReturn(TEST_EMAIL)
+        `when`(registrationParams.password).thenReturn(TEST_PASSWORD)
+        `when`(registrationParams.activationCode).thenReturn(TEST_ACTIVATION_CODE)
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCaseTest.kt:42]: you can move these to setUp() I guess, since they're same in every test method

File: app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountWithEmailViewModel.kt [hunk 2]
```diff
@@ -22,20 +26,41 @@ import kotlinx.coroutines.Dispatchers
...
+        _emailLiveData.value = email
+    }
+
+    fun saveName(name: String) {
+        _nameLiveData.value = name
+    }
+
+    fun saveActivationCode(activationCode: String) {
+        _activationCodeLiveData.value = activationCode
+    }
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountWithEmailViewModel.kt:63]: I'm guessing these values will not be stored if a process death happens, since viewmodel will be dead too, when Activity dies.
You can store them with ViewModel SavedState, or fragment arguments & savedInstanceState. You can handle that in a separate commit if you like, I'm approving this one since otherwise the PR looks good.

File: app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountNameInputFragment.kt [hunk 3]
```diff
@@ -27,6 +34,10 @@ class CreatePersonalAccountNameInputFragment : Fragment(R.layout.fragment_create
         )
     }
 
+    private fun saveName() {
+        createPersonalAccountViewModel.saveName(createPersonalAccountNameInputEditText.text.toString())
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountNameInputFragment.kt:38]: Do we not need to validate the name in anyway first before we save it? Would it be worth trimming the whitespace on the name here? Incase we get any accidental spaces at-least?

File: app/src/main/kotlin/com/waz/zclient/feature/auth/registration/register/RegisterRepository.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.waz.zclient.feature.auth.registration.register
+
+import com.waz.zclient.core.exception.Failure
+import com.waz.zclient.core.functional.Either
+
+interface RegisterRepository {
+    suspend fun registerPersonalAccountWithEmail(
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/feature/auth/registration/register/RegisterRepository.kt:7]: These parameters aren't wrapped properly, the last bracket needs to be on a new line

File: app/src/main/kotlin/com/waz/zclient/feature/auth/registration/register/datasources/remote/RegisterRemoteDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        password: String,
+        activationCode: String
+    ): Either<Failure, UserResponse> = request {
+        activationApi.register(RegisterRequestBody(
+            name = name,
+            email = email,
+            password = password,
+            emailCode = activationCode,
+            locale = Locale.getDefault().toLanguageTag(),
+            label = UUID.randomUUID().toString()
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/feature/auth/registration/register/datasources/remote/RegisterRemoteDataSource.kt:25]: I'm curious as to how you're unit testing the validity of these values?

File: app/src/main/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+class RegisterPersonalAccountWithEmailUseCase(private val registerRepository: RegisterRepository) :
+    UseCase<Unit, RegistrationParams>() {
+    override suspend fun run(params: RegistrationParams): Either<Failure, Unit> =
+        registerRepository.registerPersonalAccountWithEmail(
+            params.name,
+            params.email,
+            params.password,
+            params.activationCode
+        )
+            .fold({
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCase.kt:21]: Can be on previous line

File: app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/datasources/remote/RegisterRemoteDataSourceTest.kt [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    fun `Given registerPersonalAccountWithEmail() is called, when api response success, then return an success`() = runBlocking {
+
+        `when`(response.body()).thenReturn(userResponse)
+        `when`(response.isSuccessful).thenReturn(true)
+
+        `when`(registerApi.register(any())).thenReturn(response)
+
+        val response = registerRemoteDataSource.registerPersonalAccountWithEmail(TEST_NAME, TEST_EMAIL, TEST_PASSWORD, TEST_ACTIVATION_CODE)
+
+        verify(registerApi).register(any())
```
Comment (BradleyWire) [inline app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/datasources/remote/RegisterRemoteDataSourceTest.kt:53]: Would be good to not rely only on` any() `here. Maybe even capture the argument and validate the correct values are being passed?

File: app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/datasources/RegisterDataSourceTest.kt [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            TEST_EMAIL,
+            TEST_PASSWORD,
+            TEST_ACTIVATION_CODE
+        )
+
+        verify(registerRemoteDataSource).registerPersonalAccountWithEmail(
+            TEST_NAME,
+            TEST_EMAIL,
+            TEST_PASSWORD,
+            TEST_ACTIVATION_CODE)
```
Comment (BradleyWire) [inline app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/datasources/RegisterDataSourceTest.kt:55]: Last bracket can be wrapped here to be consistent with the other two calls prior

File: app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCaseTest.kt [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+            TEST_EMAIL,
+            TEST_PASSWORD,
+            TEST_ACTIVATION_CODE
+        )
+
+        response.isLeft shouldBe true
+
+        response.fold({
+            it shouldBe UnauthorizedEmail
+        }) { assert(false) }
```
Comment (BradleyWire) [inline app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCaseTest.kt:64]: What are we asserting is false here? I'm curious because I haven't seen this before and I genuinely don't know.

**Comment by BradleyWire** [comment]
Approving on the basis my comments are addressed in your next TODO PR as this one is already quite big.

File: app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountNameInputFragment.kt (hunk 3 above)
Comment (mejdoo) [reply app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountNameInputFragment.kt:38]: this is already mentioned in TODO

File: app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCaseTest.kt (hunk 9 above)
Comment (mejdoo) [reply app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCaseTest.kt:64]: After double checking, I realised that this assertion doesn't make any sense. The block is not even reachable because we have always a failure. In this case, I need to apply `fold` instead of `map` to get the value of the Failure.

I updated the code to :
```
 response.fold({
            it shouldBe UnauthorizedEmail
        }) { }
```

Comment (gizemb) [reply app/src/test/kotlin/com/waz/zclient/feature/auth/registration/register/usecase/RegisterPersonalAccountWithEmailUseCaseTest.kt]: "The block is not even reachable because we have always a failure". That's not always correct, because you cannot assume you're always getting a failure, since that is exactly what you're testing: "whether it gets a failure".

when it comes to removing the assertion, the execution is right, but the rationale isn't, as I mentioned above. Normally, `assert(false)` is there to ensure that if we accidentally get a success instead of error, the test case fails. Because if we remove it, and the result is a success, nothing will be executed in the second block of `fold`, and since we have no exception, assertion or whatsoever there, and the test case will result in a **false-positive**.

However, since we already have `response.isLeft` check on line 60, that assertion will fail if the response is success such a case, so it's indeed not needed in this specific example.
 to summarize:
```
        response.isLeft shouldBe true
AND
        response.fold({
            it shouldBe UnauthorizedEmail
        }) { }
```
is equivalent to

```
        response.fold({
            it shouldBe UnauthorizedEmail
        }) { assert(false) }
```

itself

File: app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountNameInputFragment.kt (hunk 3 above)
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountNameInputFragment.kt:38]: This one is on me then, as you were 😃

File: app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountWithEmailViewModel.kt (hunk 2 above)
Comment (mejdoo) [reply app/src/main/kotlin/com/waz/zclient/feature/auth/registration/personal/email/CreatePersonalAccountWithEmailViewModel.kt:63]: I will address that in a separate PR. I need check how to manage viewModel SavedState with Koin.
